### DISCUSSION TITLE
Show emojis on dumbphones

### DISCUSSION THREAD

**Discussion by guillezcurra** [tomtau/superbasic-im#18] [Ideas]
Hi! I have been searching online but I am lost. Do you have any idea on how can we display whatsapp emojis on the app? Is there any set of images that covers all emojis that we can maybe use?

Thank you!

**Comment by tomtau** [comment]
Maybe downloading openmoji PNGs and the app could be hosting https://www.npmjs.com/package/openmoji#downloads--distribution-channels / https://openmoji.org/

So the Unicode usage in a message gets replaced by an `<img src=...emoji png... />` ?

**Reply by guillezcurra** [reply]
I didn't know openmoji, it looks great, I see that they have a PNG 72x72 option which would work well, it seems that every image has, as a name, the corresponding unicode.
Do you know how emojis are represented in a whatsapp-ui message.body? Maybe its also a unicode?

**Reply by tomtau** [reply]
> Do you know how emojis are represented in a whatsapp-ui message.body? Maybe it's also a unicode?

AFAIK they are just a part of UTF-8 / Unicode string which modern browsers / OS display natively out of the box

**Reply by tomtau** [reply]
So maybe can use https://github.com/hfg-gmuend/openmoji/blob/HEAD/API.md (perhaps to make it a map instead of array with emoji strings/characters being keys ?)

**Reply by tomtau** [reply]
Or iterate the message string by characters, convert each character to its hexcode and check if there's its image file?

**Reply by guillezcurra** [reply]
Yes, maybe that's the way, but is it efficient enought? I was thinking that maybe we could do it the same way it's done in "node-emoji" unemojify function https://github.com/omnidan/node-emoji/blob/main/src/unemojify.ts but replacing the emoji for the .png file mounted as a <img> element. I think it also iterates through  characters as you suggest.

**Reply by tomtau** [reply]
Yes, node-emoji has a regex to iterate over all characters in a given Unicode string, then checks if a character is in https://www.npmjs.com/package/emojilib (which is a map from emoji characters to keywords) and if found, it replaces them with emoji names, otherwise keeps the original characters (ASCII or other Unicode).

**Reply by guillezcurra** [reply]
Should we do the same as node-emoji? But instead of returning the emoji "string" name between colons : smile : we could return an HTML element ```<img src="1F3C5.png">```, for example.

Using [openmoji API](https://github.com/hfg-gmuend/openmoji/blob/HEAD/API.md) we create a map emojis with key 'emoji' and value 'hexcode'. Then we do the e = emojis.get(emoji) and we create the ```<img>``` using this e.hexcode

**Reply by tomtau** [reply]
Yes, I think that makes sense. Just maybe need to be careful to html-escape the message content before passing it to that emoji image conversion function (so that only emoji images are interpreted as html)

**Reply by guillezcurra** [reply]
Okay, I will give it a try

**Reply by guillezcurra** [reply]
I have been setting this up and I tried to create a function emojiToImage similar to [ unemijofy.ts ](https://github.com/omnidan/node-emoji/blob/main/src/unemojify.ts) but I have an error when trying to import char-regex into our file helper.ts:
```javascript
 import  charRegex from 'char-regex';
 ```
 You can see it in this branch I am developing [helper.ts](https://github.com/guillezcurra/superbasic-im/blob/emoji_a_imatge/src/helper.ts).
 
 The error I find is shown when trying to reload pm2 restart index.
```
Error [ERR_REQUIRE_ESM]: require() of ES Module /home/guille/superbasic-im/node_modules/char-regex/index.js from /home/guille/superbasic-im/dist/helper.js not supported.
Instead change the require of index.js in /home/guille/superbasic-im/dist/helper.js to a dynamic import() which is available in all CommonJS modules.
     at Hook._require.Module.require (/usr/local/lib/node_modules/pm2/node_modules/require-in-the-middle/index.js:101:39)
     at Object.<anonymous> (/home/guille/superbasic-im/dist/helper.js:7:38)
     at Hook._require.Module.require (/usr/local/lib/node_modules/pm2/node_modules/require-in-the-middle/index.js:101:39)
     at Object.<anonymous> (/home/guille/superbasic-im/dist/routes.js:44:18)
     at Hook._require.Module.require (/usr/local/lib/node_modules/pm2/node_modules/require-in-the-middle/index.js:101:39)
    at Object.<anonymous> (/home/guille/superbasic-im/dist/index.js:52:18)
    at Object.<anonymous> (/usr/local/lib/node_modules/pm2/lib/ProcessContainerFork.js:33:23) {
  code: 'ERR_REQUIRE_ESM'
 }
```

Do you know what is the problem? I was searching a solution on the internet but couldn't find anything that works. I am stuck.

Thanks

**Reply by tomtau** [reply]
looks like some TS compiler issue, so it can be solved by tweaking tsconfig.json ... Alternatively, that dependency looks like a micropackage, so perhaps just this file can be copy-pasted: https://github.com/Richienb/char-regex/blob/main/index.js instead?

**Reply by guillezcurra** [reply]
Thank you! I copied the file as you said and now everything works, it finds the image well and everything. I have now problems with dimensions because 72px is too much. I have solved this for modern browsers using this style:
```html
style="height: 1em; width: 1em;"
```
But Opera Mini 7.1 seems to ignore it. 
Any ideas on how to make it adjust image size to match text size?
This is the problem :)
<img src="https://github.com/user-attachments/assets/053ae19f-0607-456c-af5e-f621e4fd1b94"  width="280" height="400" />

Maybe adding a fixed width and height that is a middleground for all dumphones?
You can check all code in my branch [emoji_a_imatge](https://github.com/guillezcurra/superbasic-im/tree/emoji_a_imatge) if you want
Thanks!

**Reply by guillezcurra** [reply]
I was now thinking that maybe there could be a "settings" page where you could "test" the ideal width and height for emojis on your phone and save it.

**Reply by tomtau** [reply]
the img tag should have width and height attributes: https://www.w3schools.com/tags/att_img_width.asp which hopefully even old browsers should respect. Those attributes are only in pixels though

**Reply by tomtau** [reply]
Yes, maybe emojis enable/disable and their size could be configurable. (Could be a page or just in a file/from a command-line)

**Reply by guillezcurra** [reply]
Opera Mini 7.1 doesn't display any images when trying this approach:
```html
<img src="../public/emoji/1F44D-1F3FC.svg" width="25" height="25" />
```
I see it working well in Symbian s60 built in browser and my desktop firefox (i changed from png to svg also, but tried before with png and the same result).
In this site to search compatibility [https://caniuse.com/mdn-css_properties_height](https://caniuse.com/mdn-css_properties_height) it appears as "Support unknown".

**Reply by tomtau** [reply]
That's a bummer. Maybe the last option is to resize the image files, either all of them via a CLI command/script in advance or on the fly: https://sharp.pixelplumbing.com/
But it may be an overkill.

**Reply by guillezcurra** [reply]
After a lot of debugging I have seen that it only shows images with at least width="30" height="30", if less than 30px the browser doesn't show them. This is how it looks:
<img src="https://github.com/user-attachments/assets/61a91e55-59ad-4e0a-a7d7-19aefe1a7f66" height=400 width=280>

By the way, I noticed some emojis are not found, in that case I will put unemojify also on them because an invalid emoji does weird things when rendering. And emojis will only be found in: group description, chats, replied messages and reactions. But not in the list of recent chats or in contacts.

Do you think this is okay? Or maybe we try another approach?

**Reply by tomtau** [reply]
I think that's ok (or what's a different approach? A different emoji image set?). Which emojis did you find were missing? But yes, I think it makes sense to put the textual representation on the ones that are not found (it was also one motivation why I put the unemojify originally, because I remember that Unicode did strange things in Opera Mini like displaying weird characters or cutting off remaining texts)

**Reply by guillezcurra** [reply]
It is weird, It does find the "thumbs up" emoji with white skin but doesn't find the emoji with yellow skin:
<img src="https://github.com/user-attachments/assets/4f31b126-baa1-4fc1-ae67-2a51510cc0f0" height=300 width=300>

**Reply by tomtau** [reply]
Strange, it should support different skin tones: https://openmoji.org/library/emoji-1F44D/

Isn't it that the emojis with skin tone modifiers are represented with two characters (emoji + modifier), and that regex search-based approach for conversion sequences it individually instead of together? (Not sure what the root cause is, just guessing what it could be)

**Reply by guillezcurra** [reply]
I think the problem is the opposite, I can see all tones of skin "thumbs up" except the original yellow one. But I see 1F44D.svg in my images project, I will try to debug it to see what is happening but seems to detect 1F44D-1F3FF or 1F44D-1F3FB but not he original 1F44D.

**Reply by tomtau** [reply]
Maybe one more option for emojis: just keep the original message text with unicode, but call twitter's emoji library to parse it and replace with images: https://github.com/jdecked/twemoji?tab=readme-ov-file#dom-parsing

But not sure if Opera Mini's JS engine can execute Twemoji

**Reply by guillezcurra** [reply]
If I understand correctly, this is executed from the browser after loading the page? Shouldn't we keep all processing work on the server-side to make it work in all browsers? Or maybe I don't understand how does it work

**Reply by tomtau** [reply]
> I think the problem is the opposite, I can see all tones of skin "thumbs up" except the original yellow one. But I see 1F44D.svg in my images project, I will try to debug it to see what is happening but seems to detect 1F44D-1F3FF or 1F44D-1F3FB but not he original 1F44D.

Oh I see; it could be a different reason then.

> If I understand correctly, this is executed from the browser after loading the page? Shouldn't we keep all processing work on the server-side to make it work in all browsers? Or maybe I don't understand how does it work

Yes, that's how it works. It can also work on the server-side, just need to be careful about sanitising the input text beforehand

**Reply by guillezcurra** [reply]
And what would be the size of these emojis? Does the size of the images change dynamically based on the viewport's dimensions?

**Reply by tomtau** [reply]
Not sure. I assume in those display size, they typically have 16x16 or 8x8 for a font character, so perhaps something around that for emojis?

**Reply by guillezcurra** [reply]
Yes, I have seen that the best size is something arount 25px. I will try to create anothe branch to develop the Twemoji option and see if its better than my implementation.
